### DISCUSSION TITLE
Implicit internal implementation of an internal interface

### DISCUSSION THREAD

**Discussion by Pyrdacor** [dotnet/csharplang#6472] [Q&A] [answered]
Is there a reason why internal interfaces can't be implemented by internal methods implicitly?

Example:

```cs
internal interface IFoo
{
    void Method();
}

public class Foo : IFoo
{
    internal void Method()
    {
        // CS0737
    }
}

internal class Bar : IFoo
{
    internal void Method()
    {
        // CS0737
    }
}
```

I know that interface members must be public when implementing them implicitly. But is there a logical reason for that?

I mean nothing outside the assembly can use that interface and in my understanding an internal method is just "public inside the same assembly".

I also know that I can achieve this via explicit implementation but this is really ugly when it comes to static methods as I can't use `MyDerivedClass.MyMethod` anymore and have to use a generic wrapper with type constraint.

**Comment by Joe4evr** [comment, marked as answer]
#294

**Reply by Pyrdacor** [reply]
Sorry I really used the search but must have overlooked this. Thanks for pointing me there. Still it's lacking a bit information about the reasoning. I guess this might be added in the future? Especially with static interface methods, this would be benefical imho.

**Comment by HaloFour** [comment]
The escape hatch is that you can always explicitly implement and delegate to an `internal` method:

```cs
public class Foo : IFoo
{
    void IFoo.Method() => this.Method();

    internal void Method()
    {
        // do something
    }
}
```

This is a language limitation.  The runtime actually is much more flexible about implementations and overrides, name and accessibility may be different as long as the signature is the same.  Languages like VB.NET expose this directly:

```vb
Public Class Foo
    Implements IFoo

    Friend Sub SomeOtherName() Implements IFoo.Method
        ' do something here
    End Sub
End Class
```

**Reply by Pyrdacor** [reply]
Very interesting. Thanks for your reply!

**Comment by DanTravison** [comment]
This thread appears to say "It's this way because that's the way it's defined" which seems to miss the point of the discussion. I have a set of public types that declare an internal interface. The interface provides identity and context to various layers and is virtually all property based and, except for two, which are stateful, all are implemented as auto properties. External to the type, all access is through the interface.  The issue is the actual implementation class.  I either need to cast 'this' everywhere I need to access the property, or I need to spell out the property twice, once for the interface and a second time for the class's usage or I need to define them all with a backing field, so I don't need to go through duplicate coding.  So, the problem comes up when the interface cannot be implemented using inheritance such as when the implementing classes are orthogonal or the implementation is a struct or ref struct.
